controllers/cluster_controller.go

```diff
@@ -205,6 +206,15 @@ func (r *ClusterReconciler) reconcileMetrics(_ context.Context, cluster *cluster
 func (r *ClusterReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster) (reconcile.Result, error) {
 	logger := r.Log.WithValues("cluster", cluster.Name, "namespace", cluster.Namespace)
 
+	// Delete the control plane first or else it will fight with this controller to recreate machines.
+	if cluster.Spec.ControlPlaneRef != nil {
+		err := r.deleteObjectViaReference(ctx, cluster, cluster.Spec.ControlPlaneRef)
+		// The only way to pass this is to try and get the ref object and find it is gone.
+		if !apierrors.IsNotFound(errors.Cause(err)) {
+			return ctrl.Result{}, err
+		}
+	}
```

Comment:
This seems ok to me but I have a question around cooperation between this and KCP controller.

We're doing `listDescendants`, which also returns control plane machines. Are we ok with this controller and the KCP controller both deleting the control plane nodes? Or should we only delete those if the KCP reference isn't there?

Reply:
I hear you. I'm finding these two controllers are overlapping in responsibility. I've been kind of wondering when this would happen based on the loose RBAC permissions the kubeadm control plane needs.

Anyway, I think your suggestion makes sense to me. If there is a control plane ref then we can assume it will clean up the control plane nodes. 

* If there is a control plane ref
    * delete the control-plane-ref and ignore control plane machines
* If there is no control plane ref
    * delete all machines